### PR TITLE
Fix pause-release not working on i.MX

### PR DESCRIPTION
**Pull request by LaurentiuM1234**
Currently, when we're trying to pause and then resume a stream we get an error message from the FW saying that a `comp_trigger()` operation failed. This was caused by the fact that `edma_release` would set the channel state to COMP_STATE_ACTIVE and `edma_start` would not allow the transition from that state.

In order to fix this, `edma_release` now changes the state to COMP_STATE_PRE_ACTIVE and `edma_start` allow the transition from COMP_STATE_PRE_ACTIVE to COMP_STATE_ACTIVE. This way, the EDMA states will be consistent with the comp state diagram found at https://thesofproject.github.io/latest/architectures/firmware/components/component-overview.html.

Another problem with `edma_start` was the fact that it was allowing the transition from COMP_STATE_SUSPEND to COMP_STATE_ACTIVE which is wrong because that state is never set in the edma driver.

Another solution to the pause-resume problem would be to simply set the state of the edma channel to COMP_STATE_PREPARE on `edma_release`. This would also work but it would be inconsistent with the aforementioned comp state diagram.

### CODE REVIEW

**Comment by dbaluta** [comment]
@LaurentiuM1234 very good description of the issues. Looking at the other drivers I would suggest using `COMP_STATE_PREPARE`. I think the documentation with the diagram state is outdated.

File: src/drivers/imx/edma.c [hunk 1]
```diff
@@ -147,7 +147,7 @@ static int edma_start(struct dma_chan_data *channel)
 	tr_info(&edma_tr, "EDMA: start(%d)", channel->index);
 
 	if (channel->status != COMP_STATE_PREPARE &&
-	    channel->status != COMP_STATE_SUSPEND)
+	    channel->status != COMP_STATE_PAUSED)
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: same error for imx/sdma.c

**Comment by lgirdwood** [comment]
Fwiw, @lyakh is working out how we can simplify the pipeline state machine by reducing/merging some states.

File: src/drivers/imx/edma.c (hunk 1 above)
Comment (LaurentiuM1234) [reply src/drivers/imx/edma.c:150]: Solved

**Review by dbaluta — APPROVED** [review]

**Comment by LaurentiuM1234** [comment]
Update:

After the first 3 commits, the problem with pause-release operation not working at all was fixed, but after a sequence of pause-release operations a static sound would play during the entire song. This static sound would sometimes disappear after another sequence of pause-release operations and would appear again after another sequence of those operations.

In order to fix this, I had to add a new function in the i.MX SAI driver: `sai_release()` which would handle the release operation on `sai_trigger(RELEASE)`. Initially, `release` would be treated just as `start` which is not really correct because:

-    we don't need to do a software reset of the SAI (according to the documentation, apart from resetting the SAI internal logic, it was also resetting the the FIFO read and write pointers which was not required)

-    we don't need to enable the flags that were not disabled by `sai_stop`

Apart from the static sound problem there seemed to be another one: after some pause-release operations the sound would play with higher volume in one earbud compared to the other one. This seemed to happen only during a certain of some song (the begging). By removing the unnecessary set of `CSR_START` register in `edma_start` the problem seemed to have disappeared (I did 20 tests with the beginning of the song and this problem didn't occur). Setting `CSR_START` register in `edma_start` is not necessary because we don't need to explicitly start a channel in the software. The hardware will take care of it.

**Comment by dbaluta** [comment]
@LaurentiuM1234 can you please capture your last comment inside the commit messages. People usually look at git log when investigating code not at github comments.

**Comment by LaurentiuM1234** [comment]
Update:

1) Changed some of the commit messages in order to better explain the issues the patches are meant to solve.
2) Reordered the last 3 commits to better reflect the times at which the problems were discovered.

**Comment by dbaluta** [comment]
thanks @LaurentiuM1234 this also fixes the pause/release problem on i.MX8MP for PCM playback. I will have more testing for compress on Monday then will merge the changes.

**Review by lgirdwood — COMMENTED** [review]
@LaurentiuM1234 this looks all self contained within iMX code, I'm assuming you were satisfied that there were no errors in the core state machine ?

**Comment by dbaluta** [comment]
> @LaurentiuM1234 this looks all self contained within iMX code, I'm assuming you were satisfied that there were no errors in the core state machine ?

Yes.

**Comment by lgirdwood** [comment]
@LaurentiuM1234 pls let us know when fully tested and ready for merge as still RFC. @paulstelian97 good for you too ?

**Comment by lgirdwood** [comment]
Rerun the CI - due to CI maintenance over the weekend.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by dbaluta** [comment]
@lgirdwood this is good for us. @paulstelian97 is on vacation for the next two weeks so we need an approval for you.

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**